### PR TITLE
Add input shape record for new dygraph operator

### PR DESCRIPTION
**Pull request by rainyfly**
### PR types

Others
### PR changes

Others
### Describe

1. 为新动态图的op收集input shapes, 修改后的动态图op模板生成的示例代码为
```C++
PADDLE_API Tensor matmul(const Tensor& x, const Tensor& y, bool transpose_x, bool transpose_y) {

  Backend kernel_backend = Backend::UNDEFINED;
  DataLayout kernel_layout = DataLayout::UNDEFINED;
  DataType kernel_data_type = DataType::UNDEFINED;

  if (kernel_backend == Backend::UNDEFINED
        || kernel_layout == DataLayout::UNDEFINED
        || kernel_data_type == DataType::UNDEFINED ) {
    auto kernel_key_set = ParseKernelKeyByInputArgs(x, y);
    auto kernel_key = kernel_key_set.GetHighestPriorityKernelKey();
    if (kernel_backend == Backend::UNDEFINED) {
      kernel_backend = kernel_key.backend();
    }
    if (kernel_layout == DataLayout::UNDEFINED) {
      kernel_layout = kernel_key.layout();
    }
    if (kernel_data_type == DataType::UNDEFINED) {
      kernel_data_type = kernel_key.dtype();
    }
  }

  VLOG(6) << "matmul API kernel key: [" << kernel_backend << ", " << kernel_layout << ", "<< kernel_data_type << "]";
  auto kernel_result = phi::KernelFactory::Instance().SelectKernelOrThrowError(
      "matmul", {kernel_backend, kernel_layout, kernel_data_type});
  const auto& kernel = kernel_result.kernel;
  VLOG(6) << "matmul kernel: " << kernel;
  auto* dev_ctx = GetDeviceContextByBackend(kernel_result.has_fallback_cpu ? Backend::CPU : kernel_backend);

  auto input_x = PrepareData(x, kernel.InputAt(0), {});
  auto input_y = PrepareData(y, kernel.InputAt(1), {});
  if(platform::RecordOpInfoSupplement::IsEnabled()){
     std::vector<std::pair<const char*, std::vector<phi::DDim>>> input_shapes{            
     {"x", {            
     (*input_x).dims()}},            
     {"y", {            
     (*input_y).dims()}}};
     platform::RecordOpInfoSupplement("matmul", input_shapes);
  }

  Tensor api_output;
  auto kernel_out = SetKernelOutput(kernel_backend, &api_output);
  paddle::platform::RecordEvent *infer_shape_record_event = nullptr;
  if(paddle::platform::RecordEvent::IsEnabled()){
    infer_shape_record_event = new paddle::platform::RecordEvent("matmul infer_meta", paddle::platform::TracerEventType::OperatorInner, 1);
  }
  phi::MetaTensor meta_out(kernel_out);

  phi::MatmulInferMeta(MakeMetaTensor(*input_x), MakeMetaTensor(*input_y), transpose_x, transpose_y, &meta_out);

  if(infer_shape_record_event != nullptr){
    delete infer_shape_record_event;
  }
  using kernel_signature = void(*)(const platform::DeviceContext&, const phi::DenseTensor&, const phi::DenseTensor&, bool, bool, phi::DenseTensor*);
  auto* kernel_fn = kernel.GetVariadicKernelFn<kernel_signature>();
  paddle::platform::RecordEvent* kernel_record_event = nullptr;
  if(paddle::platform::RecordEvent::IsEnabled()){
    kernel_record_event = new paddle::platform::RecordEvent("matmul compute", paddle::platform::TracerEventType::OperatorInner, 1);
  }
    (*kernel_fn)(*dev_ctx, *input_x, *input_y, transpose_x, transpose_y, kernel_out);
  if(kernel_record_event != nullptr){
    delete kernel_record_event;
  }
  if (kernel_result.has_fallback_cpu) {

    TransDataBackend(kernel_out, kernel_backend, kernel_out);
  }

  return api_output;
}
```
2. 加入接口的控制开关，控制是否收集input shapes和memory信息

### CODE REVIEW

File: paddle/fluid/platform/profiler.cc [hunk 1]
```diff
@@ -308,6 +318,33 @@ RecordOpInfoSupplement::RecordOpInfoSupplement(
...
+    const std::map<std::string, std::vector<framework::DDim>> &input_shapes) {
+  if (FLAGS_enable_host_event_recorder_hook == false) {
+    return;
+  }
+  if (IsEnabled() == false) {
+    return;
+  }
+  std::map<std::string, std::vector<framework::proto::VarType::Type>> dtypes;
+  std::vector<std::string> callstack;
+  // for(auto it=shapes.begin(); it != shapes.end(); it++){
```
Comment (JiabinYang) [inline paddle/fluid/platform/profiler.cc]: no comment

**Review by kuizhiqing — DISMISSED** [review]
LGTM

File: paddle/phi/api/yaml/generator/api_base.py [hunk 2]
```diff
@@ -709,11 +722,16 @@ def gen_kernel_code(self, kernel_name, code_indent, inplace_flag=False):
 {code_indent}      "{kernel_name}", {{kernel_backend, kernel_layout, kernel_data_type}}{cudnn_args});
 {code_indent}  const auto& kernel = kernel_result.kernel;
 {code_indent}  VLOG(6) << "{kernel_name} kernel: " << kernel;
-
 {code_indent}  auto* dev_ctx = GetDeviceContextByBackend(kernel_result.has_fallback_cpu ? Backend::CPU : kernel_backend);
+{code_indent}  std::map<std::string, std::vector<phi::DDim>> input_shapes;
```
Comment (JiabinYang) [inline paddle/phi/api/yaml/generator/api_base.py]: No extra data structure construct to make sure performance will not be affected by it

**Review by JiabinYang — COMMENTED** [review]
some comments, and better to give a generated code sample

File: paddle/phi/api/yaml/generator/api_base.py [hunk 3]
```diff
@@ -709,11 +722,16 @@ def gen_kernel_code(self, kernel_name, code_indent, inplace_flag=False):
...
+{code_indent}  paddle::platform::RecordEvent infer_shape_record_event(\"{kernel_name} infer_shape\", paddle::platform::TracerEventType::OperatorInner, 1);
```
Comment (JiabinYang) [inline paddle/phi/api/yaml/generator/api_base.py]: same

File: paddle/phi/api/yaml/generator/api_base.py [hunk 4]
```diff
@@ -709,11 +722,16 @@ def gen_kernel_code(self, kernel_name, code_indent, inplace_flag=False):
...
 {input_tensors}
 {output_create}
 {self.gene_infer_meta(kernel_output_names, code_indent)}
+{code_indent}  infer_shape_record_event.End();
```
Comment (JiabinYang) [inline paddle/phi/api/yaml/generator/api_base.py]: same

File: paddle/phi/api/yaml/generator/api_base.py (hunk 3 above)
Comment (rainyfly) [reply paddle/phi/api/yaml/generator/api_base.py]: Done

File: paddle/phi/api/yaml/generator/api_base.py (hunk 2 above)
Comment (rainyfly) [reply paddle/phi/api/yaml/generator/api_base.py]: Done

File: paddle/fluid/platform/profiler.cc (hunk 1 above)
Comment (rainyfly) [reply paddle/fluid/platform/profiler.cc]: Done

File: paddle/phi/api/yaml/generator/api_base.py (hunk 4 above)
Comment (rainyfly) [reply paddle/phi/api/yaml/generator/api_base.py]: Done

**Review by JiabinYang — DISMISSED** [review]
LGTM

**Review by zyfncg — APPROVED** [review]

**Review by jzhang533 — APPROVED** [review]
LGTM for API changes at:
https://github.com/PaddlePaddle/Paddle/pull/44999/files#diff-77f84e25f61d58a037e6646a8e74b227b4abb05762e9eddc040b47c8c4e201b1R406

**Pull request merged**